File: Problems 0000-0099/15. 3Sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        n = len(nums)
        for i in range(0, n - 2):
            if i > 0 and nums[i-1] == nums[i]:
                continue
            if nums[i] > 0: break
            l, r = i + 1, n - 1
            while l < r:
                sumVal = nums[i] + nums[l] + nums[r]
                if sumVal < 0:
                    l += 1
                elif sumVal > 0:
                    r -= 1
                else:
                    res.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l+1]:
                        l += 1
                    while l < r and nums[r] == nums[r-1]:
                        r -= 1
                    l += 1
                    r -= 1
        
        return res

File: Problems 0000-0099/15. 3Sum/solution.ts
function threeSum(nums: number[]): number[][] {
    // IMPORTANT: By default JS sorts numbers lexicographically
    // as if they were strings
    nums.sort((a, b) => a - b);

    const n = nums.length;
    const res = [];

    for (let i = 0; i < n - 2; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        if (nums[i] > 0) break;
        
        let l = i + 1;
        let r = n - 1;

        while (l < r) {
            const sumVal = nums[i] + nums[l] + nums[r];

            if (sumVal === 0) {
                res.push([nums[i], nums[l], nums[r]]);

                // Avoid duplicates
                while (l < r && nums[l] === nums[l + 1]) {
                    l++;
                }
                l++;

                while (l < r && nums[r] === nums[r - 1]) {
                    r--;
                }
                r--;
            } else if (sumVal < 0) {
                l++;
            } else {
                r--;
            }
        }
    }

    return res;
};